[ADP-500] Change addresses are not listed as soon as a transaction is no longer pending

hasufell opened the pull request:
# Issue Number

ADP-500

# Overview

- [x] created an integration test that checks the accepctance criteria and fails on the current branch
- [x] fixed the code in `knownAddresses` to also return all used change addresses

# Comments

## Questions

Do we need to communicate this to API users?

Commit:
Add integration test for ADP-500, which ought to fail now

paweljakubas commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs:
```diff
@@ -253,6 +255,59 @@ spec = describe "SHELLEY_ADDRESSES" $ do
+        -- See that a new change address is listed there
+        addrsChange <- request @[ApiAddress n] ctx
+            (Link.listAddresses @'Shelley wSrc) Default Empty
+        verify addrsChange
```
checking the number of addresses is one thing, another is to check address status of those addresses. We need to look at `Used/Unused` here and compare later

paweljakubas commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs:
```diff
@@ -253,6 +255,59 @@ spec = describe "SHELLEY_ADDRESSES" $ do
+            (Link.listAddresses @'Shelley wSrc) Default Empty
+        verify addrsChange'
+            [ expectResponseCode HTTP.status200
+            , expectListSize 31
```
Remember, that when the address is used, address pool gap invariant kicks in. We need to have consecutive  `gap` number of Unused addresses. So it is better to observe the addresses engaged and make sure when Pending -> InLedger transition happens they address status evaluates as expected.

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs:
```diff
@@ -253,6 +255,59 @@ spec = describe "SHELLEY_ADDRESSES" $ do
+            (Link.listAddresses @'Shelley wSrc) Default Empty
+        verify addrsChange
+            [ expectResponseCode HTTP.status200
+            , expectListSize 31
```
I'm guessing that 31 is because `fixtureWallet` has address pool gap = 20 and there are funds on first 10 addresses, so initially there was 30 addresses on the wallet, so the change address is generated as next one 31st. Perhaps a comment would be handy explaining this.

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs:
```diff
@@ -253,6 +255,59 @@ spec = describe "SHELLEY_ADDRESSES" $ do
+            ]
+
+        -- 5. wait for transaction to be inserted
+        eventually "Wallet balance = initPoolGap * minUTxOValue" $ do
```
I think this should be:
```suggestion
        eventually "Wallet balance = minUTxOValue" $ do
```

initPoolGap has nothing to do here I guess.

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs:
```diff
@@ -253,6 +255,59 @@ spec = describe "SHELLEY_ADDRESSES" $ do
+                (Link.getWallet @'Shelley wDest) Default Empty
+            expectField
+                (#balance . #getApiT . #available)
+                (`shouldBe` Quantity (1 * 1_000_000))
```
```suggestion
                (`shouldBe` Quantity minUTxOValue)
```

Commit:
Always show all change addresses in `knownAddresses`

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:
```diff
@@ -794,19 +794,14 @@ instance
-            nonChangeAddresses =
+            -- instead of only showing as many unused change addresses as there
+            -- are pending transactions, we just show them all
+            visibleChangeAddresses = addresses (liftPaymentAddress @n @k) (internalPool s)
```
The question is... does this break any assumptions further down the line? Is the length of `visibleNonChangeAddresses` fixed? Do we want to add info about internal vs external? Questions over questions.

KtorZ replied:
We aren't supposed to show change addresses until they've been used in a transaction... They are internal and not for users to use.

hasufell replied:
Done

hasufell commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs:
```diff
@@ -253,6 +256,54 @@ spec = describe "SHELLEY_ADDRESSES" $ do
+        -- 5. Check that change address is still there as unused
+        -- and the gap moved.
+        addrsAfter <- listAddresses @n ctx wSrc
+        verifyAddrs addrsAfter 51 11
```
Looks good. Prior to the transaction, this was `50` (total) and `10` (used).

hasufell commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs:
```diff
@@ -253,6 +257,83 @@ spec = describe "SHELLEY_ADDRESSES" $ do
+                (#balance . #getApiT . #available)
+                (`shouldBe` fmap (1 * minUTxOValue +) currentSrcWalletBalance)
+                rb
+        listAddresses @n ctx wDest >>= verifyAddrs 22 2
```
@paweljakubas this doesn't hold true, even after the patch to `knownAddresses`. We send another transaction from the destination wallet back to the source wallet. We should see one more unused address here, or is there simply no change?

paweljakubas replied:
after sending from wDest we should have external address pool like before it, ie. Used address at index =0, and Unused from 1 up to 10.So 11 addresses in the pool. Now, one of internal addresses is Used. the question is which one. If it is the first one in the internal pool, then the above should be correct expectation, ie. we should have after two pools `<>` 
```
i=0 Used -- beginning of external pool
i=1 Unused
i=2 Unused
...
i=10 Unused -- end of external pool
i=11 Used -- beginning of internal pool
i=12 Unused 
...
i=21 Unused -- end of internal pool
```
It is not like that? maybe it is worth looking at seq state and make sure

hasufell replied:
I think there's just no change generated... I changed the money that is initially sent from `wSrc` to `wDest` to `(3 * minUTxOValue)` and suddenly the test works.

paweljakubas replied:
yes! I said in the comment above. both output and change NEEDS to be > minUTxOValue and to ensure that let's have bigger input in wDest. So that input - fee = change > minUTxOValue

paweljakubas replied:
see https://github.com/input-output-hk/cardano-wallet/pull/2370#discussion_r539154549

paweljakubas commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs:
```diff
@@ -253,6 +257,83 @@ spec = describe "SHELLEY_ADDRESSES" $ do
+
+        -- 3. send a transaction from that wallet
+        addrs <- listAddresses @n ctx wDest
+        verifyAddrs 20 0 addrs
```
that's what I expect given `initPoolGap` ie., 10 Unused change addresses and 10 Unused external addresses

paweljakubas commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs:
```diff
@@ -253,6 +257,83 @@ spec = describe "SHELLEY_ADDRESSES" $ do
+        wDest <- emptyWalletWith ctx ("Wallet", "cardano-wallet", initPoolGap)
+
+        -- 2. list addresses of that wallet (for later comparison)
+        listAddresses @n ctx wSrc >>= verifyAddrs 50 10
```
I do not think that this one is needed

hasufell replied:
Why not? I'd like to compare before the transaction and after the transaction.

paweljakubas replied:
to me this is duplicated testing. Listing addresses in `wDest` already inspects this.

paweljakubas commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs:
```diff
@@ -253,6 +257,83 @@ spec = describe "SHELLEY_ADDRESSES" $ do
+
+        -- 5. Check that change address is still there as unused
+        -- and the gap moved.
+        listAddresses @n ctx wSrc >>= verifyAddrs 51 11
```
this one is not needed - it just makes this test more sophisticated

hasufell replied:
I also don't understand why not. This was the original test case that proves the change address shows up. Test number 6. is an optional addition of sending funds back to the source wallet.

paweljakubas replied:
you already show the case when listing addresses for wDest. So for me it is kinda duplication.

paweljakubas commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs:
```diff
@@ -253,6 +257,83 @@ spec = describe "SHELLEY_ADDRESSES" $ do
+        listAddresses @n ctx wSrc >>= verifyAddrs 51 11
+
+        -- 6. send another transaction from the destination wallet
+        listAddresses @n ctx wDest >>= verifyAddrs 21 1
```
this is what should be expected after money was sent to  address with index = 0. This should be address from external pool. After receiving funds this address should have flipped to Used. And because we need `initAddressGap=10` so one Unused address extended external address pool. internal (aka change) address pool stays intact, ie. 10 Unused addresses

paweljakubas commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs:
```diff
@@ -253,6 +257,83 @@ spec = describe "SHELLEY_ADDRESSES" $ do
+                "payments": [{
+                    "address": #{destination},
+                    "amount": {
+                        "quantity": #{2 * minUTxOValue},
```
maybe it is good to send 10*minUTxOValue as we risk the situation when sending back from wDest that we can have out+fee+change where change is less than minUTxOValue and hence it is not visible in listAddresses. outputs and changes are required to be > minUTxOValue!

hasufell replied:
Done

paweljakubas reviewed (approved):
The integration testing persuades me, applied fix also. Would be good also for @KtorZ to double check the fix. Also you need to soothe hlint and stylish

hasufell commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/SequentialSpec.hs:
```diff
@@ -488,9 +488,13 @@ prop_changeIsOnlyKnownAfterGeneration (intPool, extPool) =
   where
     prop_addrsNotInInternalPool addrs =
-        map (\x -> (ShowFmt x, isNothing $ fst $ lookupAddress id x intPool)) addrs
+        map (\(x, s) ->
```
So now `prop_addrsNotInInternalPool` doesn't really hold, because we have *used known internal addresses* in the return of `knownAddresses`. That's expected afais (that was the point of this PR, no?), so I adjusted that expectation.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/SequentialSpec.hs:
```diff
@@ -488,9 +488,13 @@ prop_changeIsOnlyKnownAfterGeneration (intPool, extPool) =
+        map (\(x, s) ->
+                let notInPool = isNothing $ fst $ lookupAddress id x intPool
+                    isUsed = s == Used
+                in (ShowFmt x, notInPool || isUsed))
```
:+1:

KtorZ committed:
slightly rewrite 'knownAddresses' for Sequential states

  I removed some intermediate local declaration which created more noise than readability. Also renamed some (e.g. 'edgeChangeAddresses') to be more consistent with the rest of the module.

KtorZ committed:
fix and slightly generalize 'createWallet' in the integration DSL

  'MonadCatch' is no longer after the unlift-io integration.

KtorZ committed:
rewrite 'ADDRESS_LIST_06' to make it easier to follow

  It was confusing to define 2 wallets as 'src' and 'dest' when there are transactions made from both and to both. Also, several 'magic' numbers which did not really capture
  the 'why' behind the assertion: 'assert (a == 11)' is far less descriptive than 'assert (a == originalValue + 1)'. Also, there's no need to assert balances of each wallets,
  it's not the point of the test and is just a pain to maintain when we change something with regards to fees in the test cluster. All we really care about in these tests is
  that the transaction is at some point inserted and that, as a result there's one extra change address.
